packages/protocol-contracts/contracts/ERC20Custody.sol

```diff
@@ -0,0 +1,193 @@
+// SPDX-License-Identifier: MIT
+// v1.0, 2023-01-10
+pragma solidity 0.8.7;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+/// @title ERC20Custody.
+/// @notice ERC20Custody for depositing ERC20 assets into ZetaChain and making operations with them.
+contract ERC20Custody {
+    using SafeERC20 for IERC20;
+
+    error NotWhitelisted();
+    error NotPaused();
+    error InvalidSender();
+    error InvalidTSSUpdater();
+    error ZeroAddress();
+    error IsPaused();
+    error ZetaMaxFeeExceeded();
+    error ZeroFee();
+
+    /// @notice If custody operations are paused.
+    bool public paused;
+    /// @notice TSSAddress is the TSS address collectively possessed by Zeta blockchain validators. 
+    address public TSSAddress; 
+    /// @notice Threshold Signature Scheme (TSS) [GG20] is a multi-sig ECDSA/EdDSA protocol. 
+    address public TSSAddressUpdater;
+    /// @notice Current zeta fee for depositing funds into ZetaChain.
+    uint256 public zetaFee;
+    /// @notice Maximum zeta fee for transaction.
+    uint256 immutable public zetaMaxFee;
+    /// @notice Zeta ERC20 token 
+    IERC20 immutable public zeta;
+    /// @notice Mapping of whitelisted token => true/false
+    mapping(IERC20 => bool) public whitelisted;
+    
+    event Paused(address sender);
+    event Unpaused(address sender);
+    event Whitelisted(IERC20 indexed asset);
+    event Unwhitelisted(IERC20 indexed asset);
+    event Deposited(bytes recipient, IERC20 indexed asset, uint256 amount, bytes message);
+    event Withdrawn(address indexed recipient, IERC20 indexed asset, uint256 amount);
+
+    constructor(address _TSSAddress, address _TSSAddressUpdater, uint256 _zetaFee, uint256 _zetaMaxFee, IERC20 _zeta) {       
+        TSSAddress = _TSSAddress; 
+        TSSAddressUpdater = _TSSAddressUpdater;
+        zetaFee = _zetaFee;
+        zeta = _zeta;
+        zetaMaxFee = _zetaMaxFee;
+        paused = false;
+    }
+
+    /**
+     * @dev Update the TSSAddress in case of Zeta blockchain validator nodes churn
+     * @param _address, new tss address
+     */
+    function updateTSSAddress(address _address) external {
+        if (msg.sender != TSSAddressUpdater) {
```

Comment:
this validation it's repeated several times. You can refactor with a modifier onlyTSSAddress and another onlyTSSUpdater